source/common/common/logger.cc

```diff
@@ -60,6 +76,13 @@ void DelegatingLogSink::log(const spdlog::details::log_msg& msg) {
   }
   lock.Release();
 
+  // Hold the sink mutex while performing the actual logging. This prevents the sink from being
+  // swapped during an individual log event.
+  // TODO(mattklein123): In production this lock will never be contended. In practice, thread
+  // protection is really only needed in tests. It would be nice to figure out a test-only
+  // mechanism for this that does not require extra locking that we don't explicitly need in the
+  // prod code.
+  absl::ReaderMutexLock sink_lock(&sink_mutex_);
```

Comment:
When I was doing benchmarks around ReaderMutexLock for symbol-tables, I found that the overall performance was worse than just using normal locks.

I tried this with a co-workers 72-core machine as well. I wouldn't say my tests were conclusive but it made me shy away from these if performance is the goal.

I do have an idea for test-only locks: you could use BasicLockable and have a TestOnlyMutex implementation of that, which you could pass into the instantiating the logger in test/test_runner.cc. You might need to update LoggerContext to take two mutexes rather than one, for this purpose.

Reply:
> I tried this with a co-workers 72-core machine as well. I wouldn't say my tests were conclusive but it made me shy away from these if performance is the goal.

I guess my feeling on this is that a) logging performance is not the most critical thing since this is already after log level checks and b) it seems like the read mutex should be faster here, though if it isn't that seems like a larger problem that I Would rather not tackle right now as part of this change, so my inclination is to just leave this for now as it's technically correct and we can revisit later if needed.

> It's would need careful commenting/variable-naming as a mutex lock which is a no-op in production could be dangerous if misinterpreted by a later change.

Per above since it's not 100% critical that we maximize perf in this path I think I would rather get this flake fix in and possibly try out your idea as part of the TODO later.